controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java

```diff
@@ -0,0 +1,204 @@
+    private final Change outstandingChange;
+    private final Optional<IssueId> deploymentIssueId;
+    private final Optional<IssueId> ownershipIssueId;
+    private final Optional<User> owner;
+    private final OptionalInt majorVersion;
+    private final ApplicationMetrics metrics;
+    private final Optional<String> pemDeployKey;
+    private final OptionalLong projectId;
+    private final boolean internal;
+    private final Map<InstanceName, Instance> instances;
+
+    /**
+     * Used to create a locked application
+     *
+     * @param application The application to lock.
+     * @param lock The lock for the application.
+     */
+    LockedApplication(Application application, Lock lock) {
+        this(Objects.requireNonNull(lock, "lock cannot be null"), application.id(), application.createdAt(),
+             application.deploymentSpec(), application.validationOverrides(), application.change(),
+             application.outstandingChange(), application.deploymentIssueId(), application.ownershipIssueId(),
+             application.owner(), application.majorVersion(), application.metrics(), application.pemDeployKey(),
+             application.projectId(), application.internal(), application.instances());
+    }
+
+    private LockedApplication(Lock lock, TenantAndApplicationId id, Instant createdAt, DeploymentSpec deploymentSpec,
+                              ValidationOverrides validationOverrides, Change change, Change outstandingChange,
+                              Optional<IssueId> deploymentIssueId, Optional<IssueId> ownershipIssueId, Optional<User> owner,
+                              OptionalInt majorVersion, ApplicationMetrics metrics, Optional<String> pemDeployKey,
+                              OptionalLong projectId, boolean internal, Map<InstanceName, Instance> instances) {
+        this.lock = lock;
+        this.id = id;
+        this.createdAt = createdAt;
+        this.deploymentSpec = deploymentSpec;
+        this.validationOverrides = validationOverrides;
+        this.change = change;
+        this.outstandingChange = outstandingChange;
+        this.deploymentIssueId = deploymentIssueId;
+        this.ownershipIssueId = ownershipIssueId;
+        this.owner = owner;
+        this.majorVersion = majorVersion;
+        this.metrics = metrics;
+        this.pemDeployKey = pemDeployKey;
+        this.projectId = projectId;
+        this.internal = internal;
+        this.instances = Map.copyOf(instances);
+    }
+
+    /** Returns a read-only copy of this */
+    public Application get() {
+        return new Application(id, createdAt, deploymentSpec, validationOverrides, change, outstandingChange,
+                               deploymentIssueId, ownershipIssueId, owner, majorVersion, metrics, pemDeployKey,
+                               projectId, internal, instances.values());
+    }
+
+    public LockedApplication with(InstanceName instance, UnaryOperator<Instance> modification) {
```

Comment:
Should perhaps follow the pattern of making dedicated methods for updating various parts of the `Instance`s? Let's just have this for now, though.